RxSwift/Observables/Buffer.swift

```diff
@@ -136,3 +136,131 @@ final private class BufferTimeCountSink<Element, Observer: ObserverType>
+     - returns: An observable sequence of buffers.
+     */
+    public func buffer<TriggerElement>(trigger: Observable<TriggerElement>) -> Observable<[Element]> {
+        return BufferTrigger(source: self.asObservable(), trigger: trigger)
+    }
+    
+    /**
+     Projects each element of an observable sequence info a buffer that's sent out when the trigger Observable emits a .next event or a .complete event, if the accumulated buffer's count
+     is over zero.
+     
+     A useful real-world example of this is when you have a resource that needs to be updated depending on some asynchronous tasks and you want to accumulate all of those tasks's results
+     and just PUT/PATCH the remote resource once per batch of results instead of once per result.
+     
+     - seealso: [buffer operator on reactivex.io](http://reactivex.io/documentation/operators/buffer.html)
+     
+     - parameter debounce: Amount of time to debounce the source as the boundary trigger between each window.
+     - parameter scheduler: Scheduler to run debouncing on.
+     - returns: An observable sequence of buffers.
+     */
+    public func buffer(debounce: RxTimeInterval, scheduler: SchedulerType) -> Observable<[Element]> {
+        let shared = self.share()
+        return shared.buffer(trigger: shared.debounce(debounce, scheduler: scheduler))
+    }
+}
+
+final fileprivate class BufferTrigger<Element, TriggerElement> : Producer<[Element]> {
+    
+    fileprivate let _source: Observable<Element>
+    fileprivate let _trigger: Observable<TriggerElement>
+    
+    init(source: Observable<Element>, trigger: Observable<TriggerElement>) {
+        _source = source
+        _trigger = trigger
+    }
+    
+    override func run<O : ObserverType>(_ observer: O, cancel: Cancelable) -> (sink: Disposable, subscription: Disposable) where O.Element == [Element] {
+        let sink = BufferTriggerSink(parent: self, observer: observer, cancel: cancel)
+        let subscription = sink.run()
+        return (sink: sink, subscription: subscription)
+    }
+}
+
+final fileprivate class BufferTriggerSink<Element, TriggerElement, O: ObserverType>
+    : Sink<O>
+    , LockOwnerType
+    , ObserverType
+    , SynchronizedOnType where O.Element == [Element] {
+    typealias Parent = BufferTrigger<Element, TriggerElement>
+    typealias E = Element
+    
+    private let _parent: Parent
+    
+    let lock = RecursiveLock()
+    
+    // state
+    private let _serialDisposable = SerialDisposable()
+    private var _buffer = [E]()
+    
+    init(parent: Parent, observer: O, cancel: Cancelable) {
+        _parent = parent
+        super.init(observer: observer, cancel: cancel)
+    }
+    
+    func run() -> Disposable {
+        let disposable = SingleAssignmentDisposable()
+        _serialDisposable.disposable = disposable
+        disposable.setDisposable(_parent._trigger.subscribe { (event: Event<TriggerElement>) in
```

Comment:
Don't think the type hinting is needed here
```suggestion
        disposable.setDisposable(_parent._trigger.subscribe { event in
```